### ISSUE TITLE
`camera_to_rays` and renderer inconsistencies?

### ISSUE THREAD

**Issue by jackd** [mikedh/trimesh#486]
I'm hunting a bug related to the consistency between rendered images and silhouettes of voxels but I'm at a loss. The following example using the current `release/major` branch:
* loads a mesh
* voxelizes via binvox (with `exact=True`)
* computes frustum coordinates via a slightly modified `camera_to_rays` (though the error persists using the trimesh implementation)
* uses `VoxelGrid.is_filled` to get occupancy values on the frustum coordinates
* reduces along the final dimension to get a silhouette.

Unfortunately, the silhouette does not match the image rendered via `scene.save_image`. I've either introduced a bug with the voxel stuff, failing to interpret something correctly, or there's something going on with cameras inconsistently between openGL rendering and `camera_to_rays`.

```python
import os
import trimesh
import numpy as np
from PIL import Image

from trimesh.geometry import align_vectors
from trimesh import util
from trimesh import transformations


def camera_to_rays(camera, transform):
    """
    Differences to trimesh.camera.camera_to_rays:

    * returned origin is not tiled
    * angles not returned
    * returned rays not normalized - have unit z-component before transform
    * rays point to center of pixels.
    """
    # radians of half the field of view
    half = np.radians(camera.fov / 2.0)
    # scale it down by two pixels to keep image under resolution
    # half *= (camera.resolution - 2) / camera.resolution

    bottom_right = np.sin(half)
    # move half a pixel in
    ## pixel_size = (bottom_right - top_left) / camera.resolution
    # pixel_size = bottom_right*2 / camera.resolution
    # bottom_right -= pixel_size / 2
    bottom_right *= 1 - 1. / camera.resolution
    
    top_left = -bottom_right

    xy = util.grid_linspace(
        bounds=[top_left, bottom_right], count=camera.resolution)
    # vectors = util.unitize(np.column_stack((xy, np.ones_like(xy[:, :1]))))
    vectors = np.column_stack((xy, np.ones_like(xy[:, :1])))
    # skip normalization
    # vectors = util.unitize(vectors)
    angles = np.arcsin(xy)

    # flip the camera transform to change sign of Z
    transform = np.dot(
        transform,
        align_vectors([1, 0, 0], [-1, 0, 0]))

    # apply the rotation to the direction vectors
    vectors = transformations.transform_points(
        vectors,
        transform,
        translate=False)

    # camera origin is single point, extract from transform
    origin = transformations.translation_from_matrix(transform)
    # why tile when we can broadcast/tile externally?
    # tile it into corresponding list of ray vectors
    # origins = np.ones_like(vectors) * origin

    return origin, vectors


resolution = 256
path = os.path.realpath(
    os.path.join(os.path.dirname(__file__), '..', 'models', 'bunny.ply'))

mesh = trimesh.load(path)
vox = mesh.voxelized(1./resolution, method='binvox', exact=True)
scene = mesh.scene()
camera = scene.camera

camera.fov = 60., 60.
camera.resolution = (resolution,)*2

origin, rays = camera_to_rays(scene.camera, scene.camera_transform)
rays = rays.reshape((resolution, resolution, 3))
offset = mesh.vertices - origin

# dists is vertices projected onto central ray
dists = np.dot(offset, rays[rays.shape[0]//2, rays.shape[1]//2])
closest = np.min(dists)
farthest = np.max(dists)
z = np.linspace(closest, farthest, resolution)

coords = np.expand_dims(rays, axis=-2) * np.expand_dims(z, axis=-1)
coords += origin

frust_vox_dense = vox.is_filled(coords)
# not sure why this is neccessary... but it gives close to the right answer?
frust_vox_dense = frust_vox_dense[-1::-1, -1::-1].transpose((1, 0, 2))
sil = np.any(frust_vox_dense, axis=-1)

image = np.array(Image.open(trimesh.util.wrap_as_stream(
    scene.save_image(resolution=None))))
image = image[..., :3]

def vis():
    import matplotlib.pyplot as plt
    _, (ax0, ax1, ax2) = plt.subplots(1, 3)
    ax0.imshow(image)
    ax1.imshow(sil)
    sil_image = image.copy()
    sil_image[np.logical_not(sil)] = 0
    ax2.imshow(sil_image)
    plt.show()


vis()
```

![sil](https://user-images.githubusercontent.com/659115/60744196-a7405f00-9f29-11e9-93b1-b4681b3f3ba9.png)

**Comment by mikedh** [comment]
Oh that's a tough one. I'd bet heavily on: `there's something going on with cameras inconsistently between openGL rendering and camera_to_rays`. [This looks suspicious](https://github.com/mikedh/trimesh/blob/master/trimesh/viewer/windowed.py#L438-L450) to me, like there should be more wangling to match the *exact* camera properties. What if you replace that bit with the camera's exact FOV and resolution, and ignore window size?

**Comment by jackd** [comment]
Camera resolution and window size are the same, and `fovY = self.scene.camera.fov[1]`...

**Comment by mikedh** [comment]
So if you add these asserts they never raise:
```
assert width == self.camera.resolution[0]
assert height == self.camera.resolution[1]
```

Can you [get the matrix from pyglet](https://stackoverflow.com/questions/9849374/using-glgetfloatv-to-retrieve-the-modelview-matrix-in-pyglet) and see if it looks as expected?

**Comment by jackd** [comment]
resolved in `release/major` (3.0.0 in future)
